Mirror the Learner Analytics API with Insights as auth layer

dan-f opened the pull request:
[AN-6150](https://openedx.atlassian.net/browse/AN-6150)

@dsjen

dsjen commented on analytics_dashboard/help/middleware.py:
```diff
@@ -8,6 +10,11 @@ class HelpURLMiddleware(object):
     """
 
     def process_template_response(self, _request, response):
+        # Ignore the django rest framework Response, which unfortunately trips
```
Missing an 's' ("trips" instead of "strips"). :)

dan-f replied:
Haha! whoops!

dsjen commented on analytics_dashboard/learner_analytics_api/v0/urls.py:
```diff
@@ -0,0 +1,15 @@
+from . import views
+
+
+COURSE_ID_PATTERN = r'(?P<course_id>[^/+]+[/+][^/+]+[/+][^/]+)'
```
We define the pattern https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/courses/urls.py#L9 as well.  It would be nice to be able to refactor the const out.

dan-f replied:
Thanks! I forgot to do that. Where do you think this should go? I can see it going in the `core` app or the django settings module.

dsjen replied:
:apple:  `core` makes sense to me!

dsjen commented on analytics_dashboard/learner_analytics_api/v0/utils.py:
```diff
@@ -0,0 +1,44 @@
+
+
+class TokenAuth(requests.auth.AuthBase):
+    """A reauests auth class for DRF-style token-based authentication"""
```
spelling: "reauests" :)

dsjen commented on analytics_dashboard/settings/dev.py:
```diff
@@ -93,3 +93,10 @@
 ########## END MODULE_PREVIEW
+
+########## REST FRAMEWORK CONFIGURATION
+REST_FRAMEWORK['DEFAULT_RENDERER_CLASSES'] = (
```
It looks like these settings are defined in `base.py` already.  Does it need to be defined here?

dan-f replied:
This is partially a mistake. I meant to declare the `DEFAULT_RENDERER_CLASSES` as being only `'rest_framework.renderers.JSONRenderer'` in `base.py`, but including `'rest_framework.renderers.BrowsableAPIRenderer'` in `dev.py`. What do you think? Another option is to include both in `base.py` which is the behavior on edx.org.

dsjen replied:
I see!  I think I like your initial inclination to only include `'rest_framework.renderers.BrowsableAPIRenderer'` in `dev.py`.  I'm just thinking that the "API" at this point isn't something we'd recommend for general use.

dsjen commented on analytics_dashboard/learner_analytics_api/v0/views.py:
```diff
@@ -0,0 +1,113 @@
+    Returns 404s rather than 403s when PermissionDenied exceptions are raised.
+    """
+    not_found_developer_message = ''
+    not_found_error_code = ''
```
Making these abstract properties would ensure that they'd be extended.  Perhaps at the very least, a message that these should be set in the base class would be helpful.  What do you think?

dan-f replied:
Yes, that's a good idea!

dsjen commented on analytics_dashboard/learner_analytics_api/v0/tests/test_views.py:
```diff
@@ -0,0 +1,106 @@
+        response = self.client.get('/api/learner_analytics/v0' + self.endpoint, self.required_query_params)
+        self.assert_response_equals(response, self.no_permissions_status_code)
+
+    @ddt.data((200, {'foo': 'bar'}), (400, {'foo': 'bar'}), (500, {'foo': 'bar'}))
```
It seems like the same body is being returned.  Would it make more sense to declare that in the method?

dan-f replied:
Rather, I made the bodies different.

dsjen replied:
:+1:

dsjen commented:
Nice tests!  I left a few comments, and it seems like you're on them.  LGTM! :+1:

dan-f commented:
@dsjen I addressed all your comments. For sharing the COURSE_ID_PATTERN, I went with django settings, which is what edx-platform does, just for consistency.
